Browser integration settings rework and KeePassHTTP deprecation

phoerious opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Description
<!--- Describe your changes in detail -->
This patch makes the following changes to the browser integration settings:
- Reword and reorder options to make them easier to understand and more consistent
- Move browser checkboxes to "General" tab to make them first-class options (will avoid user questions later why integration isn't working)
- Hide unimplemented settings
- Add "Browse..." button to proxy location setting

Furthermore, it deprecates KeePassHTTP and the WITH_XC_HTTP CMake flag and introduces a new WITH_XC_NETWORKING flag which separates general network access code from KeePassHTTP browser integration. Setting WITH_XC_HTTP implies WITH_XC_NETWORKING.

A CMake warning is issued when a user compiles KeePassXC with the deprecated WITH_XC_HTTP flag. Another (one-time) warning is shown to the user upon opening a database if KeePassHTTP is enabled in the settings. This warning links to a migration guide on our website which is yet to be written.

As a last change, the global message widget was redesigned a little. It now has proper margins and the warning background color is a little friendlier (although still not perfect, contrast is hard to get right with this combination of white text and blue links on yellow-ish background).

## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
KeePassXC continues to compile and run fine with all four (three) possible combinations of the two flags WITH_XC_NETWORKING and WITH_XC_HTTP.

## Screenshots (if appropriate):
![browser-settings1](https://user-images.githubusercontent.com/911270/35020841-50a524fa-fb2e-11e7-85cc-e44ac2d0e9f8.png)
![screenshot_20180117_024319](https://user-images.githubusercontent.com/911270/35021250-5882248c-fb30-11e7-9dc0-9c46db670af4.png)
![message-widget](https://user-images.githubusercontent.com/911270/35020843-50ebaf38-fb2e-11e7-9beb-307c45345622.png)

(warning background color in the last screen shot is not quite accurate)


## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation and I have updated it accordingly.

phoerious commented:
I wonder if we should also rename keepassxc-browser into KeePassXC-Browser. It feels rather weird to ship something with our branding that has a different spelling.

TheZ3ro commented on src/browser/BrowserOptionDialog.cpp:
```diff
@@ -102,3 +112,16 @@ void BrowserOptionDialog::saveSettings()
+#else
+    QString fileTypeFilter(tr("Executable Files (*.*)"));
+#endif
+    auto proxyLocation = QFileDialog::getOpenFileName(this, tr("Select custom proxy location"),
```
Please use `FileDialog::getOpenFileName()`

phoerious replied:
I must admit I forgot about that class. But I don't think I should use it here. The main purpose of FileDialog is to remember the last open/save location. We don't want that here. In fact, it would be very annoying if the next time I click "Open Database" the selected directory would be the KeePassXC install folder.

TheZ3ro replied:
You can use `FileDialog::setNextForgetDialog()` before calling `getOpenFileName()` to not save the final destination, and you can specify the `dir` argument so you won't be placed in the `LastDir`.

Using our FileDialog implementation helps with future tests (since implements `nextFileName`), fixes focus on macOS and if in the future we will modify something that will get propagated here as well

phoerious replied:
I set a few additional flags for the QFileDialog, which can't be set via the static method, so I had to change this to a custom instantiation of QFileDialog. That makes it impossible to use FileDialog here. I therefore added the extra activateWindow() fix and would suggest to postpone this to a refactor of FileDialog, during which we make it a subclass of QFileDialog.

TheZ3ro commented on src/gui/AboutDialog.cpp:
```diff
@@ -78,20 +78,20 @@ AboutDialog::AboutDialog(QWidget* parent)
+#endif
+#ifdef WITH_XC_HTTP
+    extensions += "\n- Legacy Browser Integration (KeePassHTTP)";
 #endif
```
What about `WITH_XC_NETWORKING`?

phoerious replied:
Yeah, agreed. This should be added here.

phoerious replied:
Funnily, these all use #ifdef :laughing:

TheZ3ro replied:
I saw that too after commenting about `if defined` :rofl:

droidmonkey replied:
#ifdef is much preferred, IMO. #if defined( ) is annoying to read

phoerious commented on src/browser/BrowserOptionDialog.cpp:
```diff
@@ -102,3 +112,16 @@ void BrowserOptionDialog::saveSettings()
+
+void BrowserOptionDialog::showProxyLocationFileDialog()
+{
+#ifdef Q_OS_WIN
```
Well, we have both. defined() is necessary of you combine different conditions. I can change it, though.

TheZ3ro replied:
Nah, fine by me.
But at some point we should convert every `ifdef` to `if defined` IMHO

droidmonkey commented on src/gui/KMessageWidget.cpp:
```diff
@@ -287,7 +287,7 @@ void KMessageWidget::setMessageType(KMessageWidget::MessageType type)
             break;
         case Warning:
-            bg1.setRgb(181, 102, 0); // values taken from kcolorscheme.cpp (Neutral)
+            bg1.setRgb(191, 126, 7); // values taken from kcolorscheme.cpp (Neutral)
```
I recommend the following code instead (lighter bg, dark fg) for warnings. I am also really not a fan of the gradient, its a little aggressive and out of place with the rest of the UI.

```
    QColor bg0, bg1, bg2, border;
    QColor fg = palette().light().color();
    switch (type) {
        case Positive:
            bg1.setRgb(0, 110,  40); // values taken from kcolorscheme.cpp (Positive)
            break;
        case Information:
            bg1 = palette().highlight().color();
            break;
        case Warning:
            bg1.setRgb(214, 190, 8);
            fg = Qt::black;
            break;
        case Error:
            bg1.setRgb(191, 3, 3); // values taken from kcolorscheme.cpp (Negative)
            break;
    }
    
    // Colors
    bg0 = bg1.lighter(110);
    bg2 = bg1.darker(110);
    border = darkShade(bg1);
```

phoerious replied:
I think this whole widget should get a little redesign for 2.4.0. It served us well, but it needs a refresh.

droidmonkey commented on src/gui/MainWindow.cpp:
```diff
@@ -77,7 +78,7 @@ class HttpPlugin: public ISettingsPage
     QString name() override
     {
-        return QObject::tr("Browser Integration");
+        return QObject::tr("Browser Integration (old)");
```
Recommend "Legacy Browser Integration", changing the first word makes it stand out more. We should also change the icon, I recommend a NEW icon for the new plugin, maybe [something like this](https://visualpharm.com/assets/591/Open%20In%20Browser-595b40b75ba036ed117da0be.svg).

phoerious replied:
I wanted to name it like that, but it's too long and I know it's even longer in other languages. Depending on font size and DPI settings, stuff may get cut off. Which in the other hand, is another argument in favor of putting it first, but "old browser integration" sounded a little too naive to me.
 Windows is special with its tiny font size. You could fit a novel there.

phoerious replied:
For the icon, I'd like to have a different one, too, but we don't really have one. The one you linked looks completely out of place. A new icon needs to be consistent with the rest of the Oxygen icons. Or we need a fitting new logo for KeepasXC-Browser which is not just the KeePassXC logo.

droidmonkey replied:
I meant "like" it, not exactly that icon. Here is another idea:

(Add the KeePassXC Logo in the background on the computer monitor) http://www.iconarchive.com/show/oxygen-icons-by-oxygen-icons.org/Apps-krfb-icon.html

droidmonkey commented on src/gui/MainWindow.cpp:
```diff
@@ -407,13 +411,44 @@ MainWindow::MainWindow()
+                                "This feature has been deprecated and will be removed in the future.<br>"
+                                "Please switch to keepassxc-browser instead! For help with migration,<br>"
+                                "visit our <a href=\"https://keepassxc.org/docs/keepassxc-browser-migration\">"
+                                "keepassxc-browser migration guide</a>.</p>"),
```
Remove most of the breaks in this message, it does not render well on Windows.

phoerious replied:
Same problem as above. I had it with only one break and it looked far better on Windows. But other systems, other font sizes. The text will increase the minimum window width quite significantly. On small screens it may overflow. I tried to enable word wrapping, but that kind of caused the widget layout to fall apart.

droidmonkey commented on src/gui/MainWindow.cpp:
```diff
@@ -407,13 +411,44 @@ MainWindow::MainWindow()
+                                "keepassxc-browser migration guide</a>.</p>"),
+                         MessageWidget::Warning, true, -1);
+
+    config()->set("Http/DeprecationNoticeShown", true);
```
Change this to a counter, we will want to show it at least three times. You could write "Final Notice" on the third showing if you want.

droidmonkey commented on src/http/OptionDialog.ui:
```diff
@@ -26,6 +26,13 @@
+   <item>
+    <widget class="QLabel" name="deprecationNotice">
+     <property name="text">
+      <string>&lt;p&gt;&lt;b&gt;NOTE:&lt;/b&gt; KeePassHTTP has been deprecated and will be removed in the future.&lt;br&gt;Please switch to keepassxc-browser instead! For help with migration, visit our &lt;a href=&quot;https://keepassxc.org/docs/keepassxc-browser-migration&quot;&gt;keepassxc-browser migration guide&lt;/a&gt;.&lt;/p&gt;</string>
```
This notice needs some left padding or center and bold it. Remove the "NOTE:" part. Different color?

phoerious commented:
@droidmonkey I don't really like the yellow tint you chose, it's too pale. I therefore put a little work into it myself and gave the whole widget a little visual refresh. I changed all the colors to friendlier and warmer shades, reduced gradient strength and border radius, changed the border colour to a darker tint of the background colour and composed the close icon with an alpha-masked overlay rectangle of the foreground colour.

The result looks like this:

![screenshot_20180119_200517](https://user-images.githubusercontent.com/911270/35168239-79a59f48-fd58-11e7-9fcd-fee3ab205c88.png)
![screenshot_20180119_201501](https://user-images.githubusercontent.com/911270/35168240-79c18406-fd58-11e7-8d4e-9ce6be8a86df.png)
![screenshot_20180119_203112](https://user-images.githubusercontent.com/911270/35168242-79dea4be-fd58-11e7-9f12-4bd0d3f064a5.png)

As you can see, I replaced the deprecation notice inside the KeePassHTTP settings with a message widget as well. I removed the warning icon from the "danger" warning of the advanced settings, because it didn't go well together with the almost identical background colour.

I also added a warning counter, so the global notice is shown three times now. Lastly, I renamed the KeePassHTTP settings category to "Legacy Browser Integration"
